jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+from pathlib import Path
+from typing import Optional
+from urllib.parse import ParseResult, urlparse
+
+from dockerflow import checks
+from pydantic import BaseModel, Field
+
+from jbi import bugzilla
+from jbi.environment import DLQUrl, get_settings
+
+logger = logging.getLogger(__name__)
+
+
+class PythonException(BaseModel, frozen=True):
+    type: str
+    description: str
+    details: str
+
+    @classmethod
+    def from_exc(cls, exc: Exception):
+        return PythonException(
+            type=exc.__class__.__name__,
+            description=str(exc),
+            details="".join(traceback.format_exception(exc)),
+        )
+
+
+class QueueItem(BaseModel, frozen=True):
+    """Dead Letter Queue entry."""
+
+    timestamp: datetime = Field(default_factory=lambda: datetime.now(tz=timezone.utc))
+    payload: bugzilla.WebhookRequest
+    error: Optional[PythonException] = None
+
+    @property
+    def identifier(self):
+        return f"{self.timestamp.isoformat()}-{self.payload.bug.id}-{self.payload.event.action}-{"error" if self.error else "postponed"}"
+
+
+@lru_cache(maxsize=1)
+def get_dl_queue():
+    settings = get_settings()
+    return DeadLetterQueue(settings.dl_queue_dsn)
+
+
+class QueueBackend(ABC):
+    """An interface for dead letter queues."""
+
+    @abstractmethod
+    def ping(self):
+        """Report if the queue backend is available and ready to be written to"""
+        pass
+
+    @abstractmethod
+    async def clear(self):
+        """Remove all bugs and their items from the queue"""
+        pass
+
+    @abstractmethod
+    async def put(self, item: QueueItem):
+        """Insert item into queued items for a bug, maintaining sorted order by
+        payload event time ascending
+        """
+        pass
+
+    @abstractmethod
+    async def remove(self, bug_id: int, identifier: str):
+        """Remove an item from the target bug's queue. If the item is the last
+        one for the bug, remove the bug from the queue entirely.
+        """
+        pass
+
+    @abstractmethod
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        """Retrieve all of the queue items for a specific bug, sorted in
+        ascending order by the timestamp of the payload event.
+        """
+        pass
+
+    @abstractmethod
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        """Retrive all items in the queue, grouped by bug
+
+        Returns:
+            dict[int, list[QueueItem]]: Returns a dict of
+            {bug_id: list of events}. Each list of events sorted in ascending
+            order by the timestamp of the payload event.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        pass
+
+
+class MemoryBackend(QueueBackend):
+    def __init__(self):
+        self.existing: dict[int, list[QueueItem]] = defaultdict(list)
+
+    def ping(self):
+        return True
+
+    async def clear(self):
+        logger.debug("Clearing queue")
```

Comment:
I think we want to avoid `debug()` level. In production they are very likely to be ignored. And our volume of application logging, compared to k8s etc., is ridiculously small, so we can be generous and verbose here I think